firecracker-control/local.go

```diff
@@ -81,263 +67,251 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
 	}
 
-	return cfg, nil
+	return resp, err
 }
 
-// startMachine attempts to start an instance within timeout and runs a goroutine to monitor VM exit event
-func (s *local) startMachine(ctx context.Context, id string, machine machine) error {
-	logger := log.G(ctx).WithField("vm_id", id)
-
-	logger.Info("starting VM instance")
-
-	ctx, cancel := context.WithTimeout(ctx, firecrackerStartTimeout)
-	defer cancel()
+func (s *local) GetVMInfo(requestCtx context.Context, req *proto.GetVMInfoRequest) (*proto.GetVMInfoResponse, error) {
+	client, err := s.shimFirecrackerClient(requestCtx, req.VMID)
+	if err != nil {
+		return nil, err
+	}
 
-	ns, err := namespaces.NamespaceRequired(ctx)
+	resp, err := client.GetVMInfo(requestCtx, req)
 	if err != nil {
-		return err
+		err = errors.Wrap(err, "shim client failed to get vm info")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	logger.Debugf("using namespace: %s", ns)
+	return resp, nil
+}
 
-	if err := machine.Start(ctx); err != nil {
-		logger.WithError(err).Error("failed to start the VM")
-		return err
+// SetVMMetadata sets Firecracker instance metadata
+func (s *local) SetVMMetadata(requestCtx context.Context, req *proto.SetVMMetadataRequest) (*empty.Empty, error) {
+	client, err := s.shimFirecrackerClient(requestCtx, req.VMID)
+	if err != nil {
+		return nil, err
 	}
 
-	if err := s.publisher.Publish(ctx, startEventName, &proto.VMStart{VMID: id}); err != nil {
-		logger.WithError(err).Error("failed to publish start event")
-		return err
+	resp, err := client.SetVMMetadata(requestCtx, req)
+	if err != nil {
+		err = errors.Wrap(err, "shim client failed to set vm info")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	go func() {
-		ctx := namespaces.WithNamespace(context.Background(), ns)
+	return resp, nil
+}
 
-		if err := machine.Wait(ctx); err != nil {
-			logger.WithError(err).Error("failed to wait the VM")
-		}
+func (s *local) newShim(ns, vmID, containerdAddress string, shimSocket *net.UnixListener, fcSocket *net.UnixListener) (*exec.Cmd, error) {
+	logger := s.logger.WithField("vmID", vmID)
 
-		if err := s.publisher.Publish(ctx, stopEventName, &proto.VMStop{VMID: id}); err != nil {
-			logger.WithError(err).Error("failed to publish stop event")
-		}
+	args := []string{
+		"-namespace", ns,
+		"-address", containerdAddress,
+	}
 
-		logger.Debug("removing machine")
+	cmd := exec.Command(internal.ShimBinaryName, args...)
 
-		s.vmLock.Lock()
-		delete(s.vm, id)
-		s.vmLock.Unlock()
-	}()
+	cmd.Dir = vm.ShimDir(ns, vmID).RootPath()
 
-	return nil
-}
-
-// StopVM stops running VM instance by VM ID
-func (s *local) StopVM(ctx context.Context, req *proto.StopVMRequest) (*empty.Empty, error) {
-	logger := log.G(ctx).WithField("vm_id", req.GetVMID())
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("%s=%s", internal.VMIDEnvVarKey, vmID),
+		fmt.Sprintf("%s=%s", internal.FCSocketFDEnvKey, "4")) // TODO remove after containerd is updated to expose ttrpc server to shim
```

Comment:
We do actually (as long as we trust the Go stdlib to work as advertised) because we are setting it via `ExtraFiles`: https://golang.org/pkg/os/exec/#Cmd

In theory we could calculate this number "4" from the index of where we appended the file to `ExtraFiles` which would make it robust to any other code changes. It's kind of borderline whether that extra complication is worth it given we should hopefully be able to remove this entirely after updating containerd to just expose the ttrpc server to the shim (and the fact that containerd itself just hardcodes `3` when they do this exact same thing with the shimSocket), but I'll give updating with that a shot.